sys/shell/commands/sc_gnrc_icmpv6_echo.c

```diff
@@ -0,0 +1,432 @@
+
+typedef struct {
+    gnrc_netreg_entry_t netreg;
+    xtimer_t sched_timer;
+    msg_t sched_msg;
+    ipv6_addr_t host;
+    char *hostname;
+    unsigned long num_sent, num_recv, num_rept;
+    unsigned long long tsum;
+    unsigned tmin, tmax;
+    unsigned count;
+    size_t datalen;
+    BITFIELD(cktab, CKTAB_SIZE);
+    uint32_t timeout;
+    uint32_t interval;
+    kernel_pid_t iface;
+    uint16_t id;
+    uint8_t hoplimit;
+    uint8_t pattern;
+} _ping_data_t;
+
+static void _usage(char *cmdname);
+static int _configure(int argc, char **argv, _ping_data_t *data);
+static void _pinger(_ping_data_t *data);
+static void _print_reply(_ping_data_t *data, gnrc_pktsnip_t *icmpv6,
+                         ipv6_addr_t *from, unsigned hoplimit);
+static void _handle_reply(_ping_data_t *data, gnrc_pktsnip_t *pkt);
+static int _finish(_ping_data_t *data);
+
+int _gnrc_icmpv6_ping(int argc, char **argv)
+{
+    _ping_data_t data = {
+        .netreg = GNRC_NETREG_ENTRY_INIT_PID(ICMPV6_ECHO_REP,
+                                                 sched_active_pid),
+        .count = DEFAULT_COUNT,
+        .tmin = UINT_MAX,
+        .datalen = DEFAULT_DATALEN,
+        .timeout = DEFAULT_TIMEOUT_USEC,
+        .interval = DEFAULT_INTERVAL_USEC,
+        .id = DEFAULT_ID,
+        .pattern = DEFAULT_ID,
+    };
+    int res;
+
+    if ((res = _configure(argc, argv, &data)) != 0) {
+        return res;
+    }
+    gnrc_netreg_register(GNRC_NETTYPE_ICMPV6, &data.netreg);
+    _pinger(&data);
+    do {
+        msg_t msg;
+
+        msg_receive(&msg);
+        switch (msg.type) {
+            case GNRC_NETAPI_MSG_TYPE_RCV: {
+                _handle_reply(&data, msg.content.ptr);
+                gnrc_pktbuf_release(msg.content.ptr);
+                break;
+            }
+            case _SEND_NEXT_PING:
+                _pinger(&data);
+                break;
+            case _PING_FINISH:
+                goto finish;
+            default:
+                /* requeue wrong packets */
+                msg_send(&msg, sched_active_pid);
+                break;
+        }
+    } while (data.num_recv < data.count);
+finish:
+    xtimer_remove(&data.sched_timer);
+    res = _finish(&data);
+    gnrc_netreg_unregister(GNRC_NETTYPE_ICMPV6, &data.netreg);
+    for (unsigned i = 0;
+         i < cib_avail((cib_t *)&sched_active_thread->msg_queue);
+         i++) {
+        msg_t msg;
+
+        /* remove all remaining messages (likely caused by duplicates) */
+        if ((msg_try_receive(&msg) > 0) &&
+            (msg.type == GNRC_NETAPI_MSG_TYPE_RCV) &&
+            (((gnrc_pktsnip_t *)msg.content.ptr)->type == GNRC_NETTYPE_ICMPV6)) {
+            gnrc_pktbuf_release(msg.content.ptr);
+        }
+        else {
+            /* requeue other packets */
+            msg_send(&msg, sched_active_pid);
+        }
+    }
+    return res;
+}
+
+static void _usage(char *cmdname)
+{
+    printf("%s [-c <count>] [-h] [-i <ms interval>] [-s <packetsize>]\n",
```

Comment:
If the idea is to save flash bytes, then this whole function can be written with two calls to `fputs`.

Reply:
`fputs` is not portable to our current MSP430 port:

```
$ BOARD=z1 make -C examples/gnrc_networking
[…]
/home/mlenders/Repositories/RIOT-OS/RIOT/sys/shell/commands/sc_gnrc_icmpv6_echo.c: In function ‘_usage’:
/home/mlenders/Repositories/RIOT-OS/RIOT/sys/shell/commands/sc_gnrc_icmpv6_echo.c:150:5: error: implicit declaration of function ‘fputs’ [-Werror=implicit-function-declaration]
/home/mlenders/Repositories/RIOT-OS/RIOT/sys/shell/commands/sc_gnrc_icmpv6_echo.c:150:11: error: ‘stdout’ undeclared (first use in this function)
/home/mlenders/Repositories/RIOT-OS/RIOT/sys/shell/commands/sc_gnrc_icmpv6_echo.c:150:11: note: each undeclared identifier is reported only once for each function it appears in
```

Reply:
What a pain! At least it should be possible do only one call to `puts()` by pasting all the lines together.

Reply:
Then `cmdname` needs to be constant which it isn't (or we have a weird line-break).